File: apps/minecraft/velocity/src/main/java/crabcraft/net/crabUtilities/velocity/CrabUtilitiesVelocity.java
package crabcraft.net.crabUtilities.velocity;

import com.google.inject.Inject;
import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.event.proxy.ProxyInitializeEvent;
import com.velocitypowered.api.event.proxy.ProxyShutdownEvent;
import com.velocitypowered.api.plugin.Dependency;
import com.velocitypowered.api.plugin.Plugin;
import com.velocitypowered.api.plugin.annotation.DataDirectory;
import com.velocitypowered.api.proxy.ProxyServer;
import crabcraft.net.crabUtilities.velocity.api.StatsPushSubscriber;
import crabcraft.net.crabUtilities.velocity.api.WebServer;
import crabcraft.net.crabUtilities.velocity.awards.AwardDbWriter;
import crabcraft.net.crabUtilities.velocity.awards.AwardDefinition;
import crabcraft.net.crabUtilities.velocity.awards.AwardEvaluator;
import crabcraft.net.crabUtilities.velocity.awards.AwardLoader;
import crabcraft.net.crabUtilities.velocity.awards.AwardQueryService;
import crabcraft.net.crabUtilities.velocity.awards.StatsQueryService;
import crabcraft.net.crabUtilities.velocity.advancements.AdvancementDbWriter;
import crabcraft.net.crabUtilities.velocity.advancements.AdvancementQueryService;
import crabcraft.net.crabUtilities.velocity.advancements.AdvancementRegistry;
import crabcraft.net.crabUtilities.velocity.awards.AwardSeeder;
import crabcraft.net.crabUtilities.velocity.db.PlayerSettingsRepository;
import crabcraft.net.crabUtilities.velocity.db.BingoRepository;
import crabcraft.net.crabUtilities.velocity.db.PostgresStatsWriter;
import crabcraft.net.crabUtilities.velocity.messaging.MessageManager;
import crabcraft.net.crabUtilities.velocity.messaging.SocialSpyCommand;
import crabcraft.net.crabUtilities.velocity.messaging.VelocityChatBridge;
import crabcraft.net.crabUtilities.velocity.staffchat.RedisStaffChat;
import crabcraft.net.crabUtilities.velocity.voicechat.CallCommand;
import crabcraft.net.crabUtilities.velocity.voicechat.CallManager;
import crabcraft.net.crabUtilities.velocity.voicechat.PlayerLocationTracker;
import crabcraft.net.crabUtilities.velocity.staffchat.StaffChatManager;
import crabcraft.net.crabUtilities.velocity.staffchat.StaffChatToggleCommand;
import crabcraft.net.crabUtilities.velocity.db.AltQueryService;
import crabcraft.net.crabUtilities.velocity.db.LoginStreakService;
import crabcraft.net.crabUtilities.velocity.litebans.LiteBansInfractionService;
import crabcraft.net.crabUtilities.velocity.litebans.PunishmentEventPublisher;
import crabcraft.net.crabUtilities.velocity.update.UpdateService;
import net.luckperms.api.LuckPerms;
import net.luckperms.api.LuckPermsProvider;
import org.slf4j.Logger;

import java.nio.file.Path;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

@Plugin(
        id = "crabutilities",
        name = "CrabUtilities",
        version = BuildInfo.VERSION,
        url = "https://www.crabcraft.net",
        authors = {"Max"},
        dependencies = {@Dependency(id = "litebans", optional = true)}
)
public class CrabUtilitiesVelocity {

    private final ProxyServer server;
    private final Logger logger;
    private final Path dataDirectory;

    private RedisStaffChat redisStaffChat;
    private volatile CallManager callManager;
    private PlayerLocationTracker playerLocationTracker;
    private StaffChatManager staffChatManager;
    private MessageManager messageManager;
    private VelocityChatBridge chatBridge;
    private WebServer webServer;
    private NicknameCache nicknameCache;
    private NicknameListener nicknameListener;
    private PendingJoinManager pendingJoinManager;
    private DiscordWebhook discordWebhook;
    private StatsPushSubscriber statsPushSubscriber;
    private PostgresStatsWriter pgWriter;
    private AwardEvaluator awardEvaluator;
    private AwardDbWriter awardDbWriter;
    private AwardQueryService awardQueryService;
    private StatsQueryService statsQueryService;
    private AdvancementDbWriter advancementDbWriter;
    private AdvancementQueryService advancementQueryService;
    private VelocityConfig config;
    private UpdateService updateService;
    private AltQueryService altQueryService;
    private LoginStreakService loginStreakService;
    private LoginStreakPublisher loginStreakPublisher;
    private ConnectionListener connectionListener;
    private PlayerSettingsService playerSettingsService;
    private LiteBansInfractionService liteBansInfractionService;
    private PunishmentEventPublisher punishmentEventPublisher;
    private LuckPerms luckPerms;
    private volatile ExecutorService databaseExecutor;
    private final Object lifecycleLock = new Object();

    @Inject
    public CrabUtilitiesVelocity(ProxyServer server, Logger logger,
                                  @DataDirectory Path dataDirectory) {
        this.server = server;
        this.logger = logger;
        this.dataDirectory = dataDirectory;
    }

    @Subscribe
    public void onProxyInitialize(ProxyInitializeEvent event) {
        this.config = VelocityConfig.load(dataDirectory, logger);
        this.nicknameCache = new NicknameCache();
        this.pendingJoinManager = new PendingJoinManager();
        this.messageManager = new MessageManager(this);

        this.pgWriter = new PostgresStatsWriter(
            config.getDbUrl(), config.getDbUsername(), config.getDbPassword(), logger
        );
        initialiseDatabaseServices(config);

        try {
            this.luckPerms = LuckPermsProvider.get();
            logger.info("LuckPerms API connected.");
        } catch (IllegalStateException e) {
            logger.warn("LuckPerms not available — alt whitelist checks disabled.", e);
            this.luckPerms = null;
        }

        startRuntimeConsumers(config);
        this.chatBridge = new VelocityChatBridge(this);
        chatBridge.start();

        StaffChatToggleCommand.register(this);
        SocialSpyCommand.register(this);
        ReloadCommand.register(this);
        CallCommand.register(this);

        this.connectionListener = new ConnectionListener(this);
        server.getEventManager().register(this, connectionListener);

        if (playerLocationTracker != null) {
            logger.info("Voice cross-server location tracker enabled.");
        }

        logger.info("CrabUtilities Velocity enabled.");
    }

    @Subscribe
    public void onProxyShutdown(ProxyShutdownEvent event) {
        synchronized (lifecycleLock) {
            if (chatBridge != null) {
                chatBridge.shutdown();
                chatBridge = null;
            }
            if (connectionListener != null) {
                connectionListener.shutdown();
            }
            stopRuntimeConsumers();
            shutdownDatabaseExecutor("shutdown");
            if (pgWriter != null) {
                pgWriter.close();
                pgWriter = null;
            }
        }
        logger.info("CrabUtilities Velocity disabled.");
    }

    public void reload() {
        synchronized (lifecycleLock) {
            VelocityConfig newConfig = VelocityConfig.load(dataDirectory, logger);

            stopRuntimeConsumers();
            shutdownDatabaseExecutor("reload");

            PostgresStatsWriter oldPgWriter = this.pgWriter;
            PostgresStatsWriter newPgWriter = new PostgresStatsWriter(
                newConfig.getDbUrl(), newConfig.getDbUsername(), newConfig.getDbPassword(), logger
            );

            this.config = newConfig;
            this.pgWriter = newPgWriter;
            initialiseDatabaseServices(newConfig);

            if (oldPgWriter != null) {
                oldPgWriter.close();
            }

            startRuntimeConsumers(newConfig);

            logger.info("CrabUtilities Velocity reloaded.");
        }
    }

    private void initialiseDatabaseServices(VelocityConfig config) {
        var dataSource = pgWriter.getDataSource();
        AwardSeeder.seedIfEmpty(dataSource, logger);
        Map<String, AwardDefinition> awards = AwardLoader.loadAll(dataSource, logger);
        logger.info("Loaded {} award definitions from database", awards.size());

        this.awardEvaluator = new AwardEvaluator(awards);
        this.awardDbWriter = new AwardDbWriter(dataSource, logger);
        this.awardQueryService = new AwardQueryService(dataSource, logger);
        this.statsQueryService = new StatsQueryService(dataSource, logger);
        this.advancementDbWriter = new AdvancementDbWriter(dataSource, logger);
        AdvancementRegistry advancementRegistry = new AdvancementRegistry(logger);
        this.advancementQueryService = new AdvancementQueryService(
                dataSource, logger, advancementRegistry);
        this.altQueryService = new AltQueryService(dataSource, logger);
        this.loginStreakService = new LoginStreakService(
                dataSource, logger,
                config.getLoginStreakResetHourUtc(),
                config.getLoginStreakRequiredPlaySeconds());
        new BingoRepository(dataSource, logger);
        this.liteBansInfractionService = new LiteBansInfractionService(logger);
    }

    private void startRuntimeConsumers(VelocityConfig config) {
        this.databaseExecutor = createDatabaseExecutor();
        this.discordWebhook = new DiscordWebhook(config.getDiscordWebhookUrl(), logger);
        DiscordWebhook staffChatWebhook =
                new DiscordWebhook(config.getStaffChatDiscordWebhookUrl(), logger);

        this.loginStreakPublisher = new LoginStreakPublisher(this, config);
        this.punishmentEventPublisher = new PunishmentEventPublisher(this, config);
        this.punishmentEventPublisher.start();

        PlayerSettingsRepository settingsRepository =
                new PlayerSettingsRepository(pgWriter.getDataSource(), logger);
        this.playerSettingsService = new PlayerSettingsService(this, settingsRepository, config);
        this.playerSettingsService.start();

        this.nicknameListener = new NicknameListener(this, config);
        this.nicknameListener.start();
        server.getEventManager().register(this, nicknameListener);

        this.statsPushSubscriber = new StatsPushSubscriber(this, config, logger);
        this.statsPushSubscriber.start();

        this.webServer = new WebServer(this, config.getApiPort());
        this.webServer.start();

        this.redisStaffChat = new RedisStaffChat(this, config);
        this.staffChatManager = new StaffChatManager(this, redisStaffChat,
                staffChatWebhook, config.getStaffChatDiscordAvatarUrl());
        this.redisStaffChat.start();

        this.updateService = new UpdateService(this);
        if (config.isUpdateEnabled()) {
            updateService.start();
        }

        if (config.isVoicechatCrossServerEnabled()) {
            this.playerLocationTracker = new PlayerLocationTracker(this, config);
            this.playerLocationTracker.start();
            server.getEventManager().register(this, playerLocationTracker);

            this.callManager = new CallManager(this, config, playerLocationTracker);
            this.callManager.start();
            server.getEventManager().register(this, callManager);
        }
    }

    public ProxyServer getServer() { return server; }
    public Logger getLogger() { return logger; }
    public Path getDataDirectory() { return dataDirectory; }
    public StaffChatManager getStaffChatManager() { return staffChatManager; }
    public MessageManager getMessageManager() { return messageManager; }
    public VelocityChatBridge getChatBridge() { return chatBridge; }
    public NicknameCache getNicknameCache() { return nicknameCache; }
    public NicknameListener getNicknameListener() { return nicknameListener; }
    public PendingJoinManager getPendingJoinManager() { return pendingJoinManager; }
    public DiscordWebhook getDiscordWebhook() { return discordWebhook; }
    public WebServer getWebServer() { return webServer; }
    public PostgresStatsWriter getPgWriter() { return pgWriter; }
    public AwardEvaluator getAwardEvaluator() { return awardEvaluator; }
    public AwardDbWriter getAwardDbWriter() { return awardDbWriter; }
    public AwardQueryService getAwardQueryService() { return awardQueryService; }
    public StatsQueryService getStatsQueryService() { return statsQueryService; }
    public AdvancementDbWriter getAdvancementDbWriter() { return advancementDbWriter; }
    public AdvancementQueryService getAdvancementQueryService() { return advancementQueryService; }
    public VelocityConfig getConfig() { return config; }
    public UpdateService getUpdateService() { return updateService; }
    public AltQueryService getAltQueryService() { return altQueryService; }
    public LoginStreakService getLoginStreakService() { return loginStreakService; }
    public LoginStreakPublisher getLoginStreakPublisher() { return loginStreakPublisher; }
    public PlayerSettingsService getPlayerSettingsService() { return playerSettingsService; }
    public LiteBansInfractionService getLiteBansInfractionService() { return liteBansInfractionService; }
    public LuckPerms getLuckPerms() { return luckPerms; }
    public CallManager getCallManager() { return callManager; }

    public boolean runDatabaseTask(String taskName, Runnable task) {
        ExecutorService executor = databaseExecutor;
        if (executor == null || executor.isShutdown()) {
            logger.warn("Skipping database task {} because the executor is stopped", taskName);
            return false;
        }
        try {
            executor.execute(() -> {
                try {
                    task.run();
                } catch (Exception e) {
                    logger.error("Database task {} failed", taskName, e);
                }
            });
            return true;
        } catch (RejectedExecutionException e) {
            logger.warn("Skipping database task {} because the executor queue is full", taskName);
            return false;
        }
    }

    private void stopRuntimeConsumers() {
        CallManager calls = callManager;
        callManager = null;
        if (calls != null) {
            calls.shutdown();
            server.getEventManager().unregisterListener(this, calls);
        }
        if (statsPushSubscriber != null) {
            statsPushSubscriber.shutdown();
            statsPushSubscriber = null;
        }
        if (webServer != null) {
            webServer.stop();
            webServer = null;
        }
        if (redisStaffChat != null) {
            redisStaffChat.shutdown();
            redisStaffChat = null;
        }
        if (nicknameListener != null) {
            nicknameListener.shutdown();
            server.getEventManager().unregisterListener(this, nicknameListener);
            nicknameListener = null;
        }
        if (playerLocationTracker != null) {
            playerLocationTracker.shutdown();
            server.getEventManager().unregisterListener(this, playerLocationTracker);
            playerLocationTracker = null;
        }
        if (punishmentEventPublisher != null) {
            punishmentEventPublisher.shutdown();
            punishmentEventPublisher = null;
        }
        if (updateService != null) {
            updateService.shutdown();
            updateService = null;
        }
        if (playerSettingsService != null) {
            playerSettingsService.shutdown();
            playerSettingsService = null;
        }
        if (loginStreakPublisher != null) {
            loginStreakPublisher.shutdown();
            loginStreakPublisher = null;
        }
    }

    private void shutdownDatabaseExecutor(String reason) {
        ExecutorService executor = databaseExecutor;
        databaseExecutor = null;
        if (executor == null) return;

        executor.shutdown();
        try {
            if (!executor.awaitTermination(15, TimeUnit.SECONDS)) {
                logger.warn("Database executor did not stop cleanly during {}; interrupting queued work", reason);
                executor.shutdownNow();
                if (!executor.awaitTermination(5, TimeUnit.SECONDS)) {
                    logger.warn("Database executor still has running work after {}", reason);
                }
            }
        } catch (InterruptedException e) {
            executor.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }

    private static ExecutorService createDatabaseExecutor() {
        return new ThreadPoolExecutor(
                4, 4,
                30L, TimeUnit.SECONDS,
                new LinkedBlockingQueue<>(256),
                Thread.ofPlatform().daemon().name("CrabUtilities-DB-", 1).factory(),
                new ThreadPoolExecutor.AbortPolicy());
    }
}

File: apps/minecraft/velocity/src/main/kotlin/crabcraft/net/crabUtilities/velocity/CrabUtilitiesVelocity.kt
package crabcraft.net.crabUtilities.velocity

import com.google.inject.Inject
import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.proxy.ProxyInitializeEvent
import com.velocitypowered.api.event.proxy.ProxyShutdownEvent
import com.velocitypowered.api.plugin.Dependency
import com.velocitypowered.api.plugin.Plugin
import com.velocitypowered.api.plugin.annotation.DataDirectory
import com.velocitypowered.api.proxy.ProxyServer
import crabcraft.net.crabUtilities.velocity.advancements.AdvancementDbWriter
import crabcraft.net.crabUtilities.velocity.advancements.AdvancementQueryService
import crabcraft.net.crabUtilities.velocity.advancements.AdvancementRegistry
import crabcraft.net.crabUtilities.velocity.api.StatsPushSubscriber
import crabcraft.net.crabUtilities.velocity.api.WebServer
import crabcraft.net.crabUtilities.velocity.awards.AwardDbWriter
import crabcraft.net.crabUtilities.velocity.awards.AwardEvaluator
import crabcraft.net.crabUtilities.velocity.awards.AwardLoader
import crabcraft.net.crabUtilities.velocity.awards.AwardQueryService
import crabcraft.net.crabUtilities.velocity.awards.AwardSeeder
import crabcraft.net.crabUtilities.velocity.awards.StatsQueryService
import crabcraft.net.crabUtilities.velocity.db.AltQueryService
import crabcraft.net.crabUtilities.velocity.db.BingoRepository
import crabcraft.net.crabUtilities.velocity.db.HalloweenRepository
import crabcraft.net.crabUtilities.velocity.db.LoginStreakService
import crabcraft.net.crabUtilities.velocity.db.PlayerSettingsRepository
import crabcraft.net.crabUtilities.velocity.db.PostgresStatsWriter
import crabcraft.net.crabUtilities.velocity.litebans.LiteBansInfractionService
import crabcraft.net.crabUtilities.velocity.litebans.PunishmentEventPublisher
import crabcraft.net.crabUtilities.velocity.messaging.MessageManager
import crabcraft.net.crabUtilities.velocity.messaging.MsgCommand
import crabcraft.net.crabUtilities.velocity.messaging.SocialSpyCommand
import crabcraft.net.crabUtilities.velocity.messaging.VelocityChatBridge
import crabcraft.net.crabUtilities.velocity.staffchat.RedisStaffChat
import crabcraft.net.crabUtilities.velocity.staffchat.StaffChatManager
import crabcraft.net.crabUtilities.velocity.staffchat.StaffChatToggleCommand
import crabcraft.net.crabUtilities.velocity.update.UpdateService
import crabcraft.net.crabUtilities.velocity.voicechat.CallCommand
import crabcraft.net.crabUtilities.velocity.voicechat.CallManager
import crabcraft.net.crabUtilities.velocity.voicechat.PlayerLocationTracker
import java.nio.file.Path
import java.util.concurrent.ExecutorService
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit
import net.luckperms.api.LuckPerms
import net.luckperms.api.LuckPermsProvider
import org.slf4j.Logger

@Plugin(
    id = "crabutilities",
    name = "CrabUtilities",
    version = BuildInfo.VERSION,
    url = "https://www.crabcraft.net",
    authors = ["Max"],
    dependencies = [Dependency(id = "litebans", optional = true)],
)
open class CrabUtilitiesVelocity
@Inject
constructor(
    private val server: ProxyServer,
    private val logger: Logger,
    @DataDirectory private val dataDirectory: Path,
) {
    private var redisStaffChat: RedisStaffChat? = null
    @Volatile private var callManager: CallManager? = null
    private var playerLocationTracker: PlayerLocationTracker? = null
    private var staffChatManager: StaffChatManager? = null
    private lateinit var messageManager: MessageManager
    private var chatBridge: VelocityChatBridge? = null
    private var webServer: WebServer? = null
    private lateinit var nicknameCache: NicknameCache
    private var nicknameListener: NicknameListener? = null
    private lateinit var pendingJoinManager: PendingJoinManager
    private lateinit var discordWebhook: DiscordWebhook
    private var statsPushSubscriber: StatsPushSubscriber? = null
    private var pgWriter: PostgresStatsWriter? = null
    private lateinit var awardEvaluator: AwardEvaluator
    private lateinit var awardDbWriter: AwardDbWriter
    private lateinit var awardQueryService: AwardQueryService
    private lateinit var statsQueryService: StatsQueryService
    private lateinit var advancementDbWriter: AdvancementDbWriter
    private lateinit var advancementQueryService: AdvancementQueryService
    private lateinit var config: VelocityConfig
    private var updateService: UpdateService? = null
    private lateinit var altQueryService: AltQueryService
    private lateinit var loginStreakService: LoginStreakService
    private var loginStreakPublisher: LoginStreakPublisher? = null
    private var connectionListener: ConnectionListener? = null
    private var vanishManager: VanishManager? = null
    private var playerSettingsService: PlayerSettingsService? = null
    private var liteBansInfractionService: LiteBansInfractionService? = null
    private var punishmentEventPublisher: PunishmentEventPublisher? = null
    private var luckPerms: LuckPerms? = null
    @Volatile private var databaseExecutor: ExecutorService? = null
    private val lifecycleLock = Any()

    @Subscribe
    open fun onProxyInitialize(event: ProxyInitializeEvent) {
        config = VelocityConfig.load(dataDirectory, logger)
        nicknameCache = NicknameCache()
        pendingJoinManager = PendingJoinManager()
        messageManager = MessageManager(this)
        vanishManager = VanishManager(this)
        vanishManager!!.start()
        pgWriter = PostgresStatsWriter(config.getDbUrl(), config.getDbUsername(), config.getDbPassword(), logger)
        initialiseDatabaseServices(config)
        try {
            luckPerms = LuckPermsProvider.get()
            logger.info("LuckPerms API connected.")
        } catch (e: IllegalStateException) {
            logger.warn("LuckPerms not available — alt whitelist checks disabled.", e)
            luckPerms = null
        }
        startRuntimeConsumers(config)
        chatBridge = VelocityChatBridge(this)
        chatBridge!!.start()
        MsgCommand.register(this)
        StaffChatToggleCommand.register(this)
        SocialSpyCommand.register(this)
        ReloadCommand.register(this)
        CallCommand.register(this)
        connectionListener = ConnectionListener(this)
        server.eventManager.register(this, connectionListener!!)
        if (playerLocationTracker != null) logger.info("Voice cross-server location tracker enabled.")
        logger.info("CrabUtilities Velocity enabled.")
    }

    @Subscribe
    open fun onProxyShutdown(event: ProxyShutdownEvent) {
        synchronized(lifecycleLock) {
            chatBridge?.shutdown()
            chatBridge = null
            connectionListener?.shutdown()
            stopRuntimeConsumers()
            vanishManager?.shutdown()
            vanishManager = null
            shutdownDatabaseExecutor("shutdown")
            pgWriter?.close()
            pgWriter = null
        }
        logger.info("CrabUtilities Velocity disabled.")
    }

    open fun reload() {
        synchronized(lifecycleLock) {
            val newConfig = VelocityConfig.load(dataDirectory, logger)
            stopRuntimeConsumers()
            shutdownDatabaseExecutor("reload")
            val oldPgWriter = pgWriter
            val newPgWriter =
                PostgresStatsWriter(newConfig.getDbUrl(), newConfig.getDbUsername(), newConfig.getDbPassword(), logger)
            config = newConfig
            pgWriter = newPgWriter
            initialiseDatabaseServices(newConfig)
            oldPgWriter?.close()
            startRuntimeConsumers(newConfig)
            logger.info("CrabUtilities Velocity reloaded.")
        }
    }

    private fun initialiseDatabaseServices(config: VelocityConfig) {
        val dataSource = pgWriter!!.getDataSource()
        AwardSeeder.seedIfEmpty(dataSource, logger)
        val awards = AwardLoader.loadAll(dataSource, logger)
        logger.info("Loaded {} award definitions from database", awards.size)
        awardEvaluator = AwardEvaluator(awards)
        awardDbWriter = AwardDbWriter(dataSource, logger)
        awardQueryService = AwardQueryService(dataSource, logger)
        statsQueryService = StatsQueryService(dataSource, logger)
        advancementDbWriter = AdvancementDbWriter(dataSource, logger)
        val advancementRegistry = AdvancementRegistry(logger)
        advancementQueryService = AdvancementQueryService(dataSource, logger, advancementRegistry)
        altQueryService = AltQueryService(dataSource, logger)
        loginStreakService =
            LoginStreakService(
                dataSource,
                logger,
                config.getLoginStreakResetHourUtc(),
                config.getLoginStreakRequiredPlaySeconds(),
            )
        BingoRepository(dataSource, logger)
        HalloweenRepository(dataSource, logger)
        liteBansInfractionService = LiteBansInfractionService(logger)
    }

    private fun startRuntimeConsumers(config: VelocityConfig) {
        databaseExecutor = createDatabaseExecutor()
        discordWebhook = DiscordWebhook(config.getDiscordWebhookUrl(), logger)
        val staffChatWebhook = DiscordWebhook(config.getStaffChatDiscordWebhookUrl(), logger)
        loginStreakPublisher = LoginStreakPublisher(this, config)
        punishmentEventPublisher = PunishmentEventPublisher(this, config)
        punishmentEventPublisher!!.start()
        val settingsRepository = PlayerSettingsRepository(pgWriter!!.getDataSource(), logger)
        playerSettingsService = PlayerSettingsService(this, settingsRepository, config)
        playerSettingsService!!.start()
        nicknameListener = NicknameListener(this, config)
        nicknameListener!!.start()
        server.eventManager.register(this, nicknameListener!!)
        statsPushSubscriber = StatsPushSubscriber(this, config, logger)
        statsPushSubscriber!!.start()
        webServer = WebServer(this, config.getApiPort())
        webServer!!.start()
        val redis = RedisStaffChat(this, config)
        redisStaffChat = redis
        staffChatManager = StaffChatManager(this, redis, staffChatWebhook, config.getStaffChatDiscordAvatarUrl())
        redis.start()
        updateService = UpdateService(this)
        if (config.isUpdateEnabled()) updateService!!.start()
        if (config.isVoicechatCrossServerEnabled()) {
            val tracker = PlayerLocationTracker(this, config)
            playerLocationTracker = tracker
            tracker.start()
            server.eventManager.register(this, tracker)
            val calls = CallManager(this, config, tracker)
            callManager = calls
            calls.start()
            server.eventManager.register(this, calls)
        }
    }

    open fun getServer() = server

    open fun getLogger() = logger

    open fun getDataDirectory() = dataDirectory

    open fun getStaffChatManager(): StaffChatManager? = staffChatManager

    open fun getMessageManager(): MessageManager? = if (::messageManager.isInitialized) messageManager else null

    open fun getChatBridge(): VelocityChatBridge? = chatBridge

    open fun getNicknameCache() = nicknameCache

    open fun getNicknameListener(): NicknameListener? = nicknameListener

    open fun getPendingJoinManager() = pendingJoinManager

    open fun getDiscordWebhook() = discordWebhook

    open fun getWebServer(): WebServer? = webServer

    open fun getPgWriter(): PostgresStatsWriter? = pgWriter

    open fun getAwardEvaluator(): AwardEvaluator? = if (::awardEvaluator.isInitialized) awardEvaluator else null

    open fun getAwardDbWriter(): AwardDbWriter? = if (::awardDbWriter.isInitialized) awardDbWriter else null

    open fun getAwardQueryService(): AwardQueryService? =
        if (::awardQueryService.isInitialized) awardQueryService else null

    open fun getStatsQueryService() = statsQueryService

    open fun getAdvancementDbWriter(): AdvancementDbWriter? =
        if (::advancementDbWriter.isInitialized) advancementDbWriter else null

    open fun getAdvancementQueryService() = advancementQueryService

    open fun getConfig() = config

    open fun getUpdateService(): UpdateService? = updateService

    open fun getAltQueryService(): AltQueryService? = if (::altQueryService.isInitialized) altQueryService else null

    open fun getLoginStreakService(): LoginStreakService? =
        if (::loginStreakService.isInitialized) loginStreakService else null

    open fun getLoginStreakPublisher(): LoginStreakPublisher? = loginStreakPublisher

    open fun getPlayerSettingsService(): PlayerSettingsService? = playerSettingsService

    open fun getVanishManager(): VanishManager = vanishManager!!

    open fun getLiteBansInfractionService(): LiteBansInfractionService? = liteBansInfractionService

    open fun getLuckPerms(): LuckPerms? = luckPerms

    open fun getCallManager(): CallManager? = callManager

    open fun runDatabaseTask(taskName: String, task: Runnable): Boolean {
        val executor = databaseExecutor
        if (executor == null || executor.isShutdown) {
            logger.warn("Skipping database task {} because the executor is stopped", taskName)
            return false
        }
        try {
            executor.execute {
                try {
                    task.run()
                } catch (e: Exception) {
                    logger.error("Database task {} failed", taskName, e)
                }
            }
            return true
        } catch (_: RejectedExecutionException) {
            logger.warn("Skipping database task {} because the executor queue is full", taskName)
            return false
        }
    }

    private fun stopRuntimeConsumers() {
        val calls = callManager
        callManager = null
        if (calls != null) {
            calls.shutdown()
            server.eventManager.unregisterListener(this, calls)
        }
        statsPushSubscriber?.shutdown()
        statsPushSubscriber = null
        webServer?.stop()
        webServer = null
        redisStaffChat?.shutdown()
        redisStaffChat = null
        nicknameListener?.let {
            it.shutdown()
            server.eventManager.unregisterListener(this, it)
        }
        nicknameListener = null
        playerLocationTracker?.let {
            it.shutdown()
            server.eventManager.unregisterListener(this, it)
        }
        playerLocationTracker = null
        punishmentEventPublisher?.shutdown()
        punishmentEventPublisher = null
        updateService?.shutdown()
        updateService = null
        playerSettingsService?.shutdown()
        playerSettingsService = null
        loginStreakPublisher?.shutdown()
        loginStreakPublisher = null
    }

    private fun shutdownDatabaseExecutor(reason: String) {
        val executor = databaseExecutor
        databaseExecutor = null
        if (executor == null) return
        executor.shutdown()
        try {
            if (!executor.awaitTermination(15, TimeUnit.SECONDS)) {
                logger.warn("Database executor did not stop cleanly during {}; interrupting queued work", reason)
                executor.shutdownNow()
                if (!executor.awaitTermination(5, TimeUnit.SECONDS))
                    logger.warn("Database executor still has running work after {}", reason)
            }
        } catch (_: InterruptedException) {
            executor.shutdownNow()
            Thread.currentThread().interrupt()
        }
    }

    companion object {
        private fun createDatabaseExecutor(): ExecutorService =
            ThreadPoolExecutor(
                4,
                4,
                30L,
                TimeUnit.SECONDS,
                LinkedBlockingQueue(256),
                Thread.ofPlatform().daemon().name("CrabUtilities-DB-", 1).factory(),
                ThreadPoolExecutor.AbortPolicy(),
            )
    }
}
